### ISSUE TITLE
TF window opening outside the display

### ISSUE THREAD

**Issue by Procyon-b** [cxw42/TabFern#189]
I recall reading another issue mentioning this bug, but I can't find it.  (**Edit from cxw** #18 and #141 are closest; also related to #111)

I'm running chrome with TF on a small screen (1280x800). It works well for everything but all windows are maximized (not TF obviously). It's probably the reason for this bug to happen: I have used TF on another computer with a much bigger display, chrome not maximized, and never TF out of sight.

```js
chrome.storage.local.get( "tabfern-window-location", function(items){
  console.info(items);
  console.info( JSON.stringify(items) );
});
```
Gives always the same value in case of error:
`{"tabfern-window-location":{"height":200,"left":-32000,"top":-32000,"width":160}}`

To work around this problem, I have devised these scriplets:
```js
chrome.storage.local.get( function(items){
  // create a backup copy
  chrome.storage.local.set(
    {"tabfern-window-location-bak": items["tabfern-window-location"] }
    );
});

// this one to restore the saved values:
chrome.storage.local.get( function(items){
  chrome.storage.local.set(
    {"tabfern-window-location": items["tabfern-window-location-bak"] }
    );
});
```
I  have used scriptlet 1 once, and since then I use scriplet 2 to "reset" TF window:
focus TF window (even if not visble), F12 to open console, run script 2, close devtools, F5 (reload TF window)

**Comment by cxw42** [comment]
Yes, there is another issue open - I'll find it.  You can also double-click the TF icon in any Chrome window to summon the TF window to that Chrome window.  Position backup is an interesting idea - we could theoretically support any number of saved window positions.

**Comment by Procyon-b** [comment]
>You can also double-click the TF icon in any Chrome window to summon the TF window

But the size and position wouldn't be identical to the previous saved infos (IIRC).

>we could theoretically support any number of saved window positions.

exactly what I though

**Comment by cxw42** [comment]
Summon changes position but not size, as I recall.  Anyway, PRs welcome! :)

**Comment by Procyon-b** [comment]
> Anyway, PRs welcome! :)

;)

I'll look into it, and see what I can do.

**Comment by Procyon-b** [comment]
> Summon changes position but not size, as I recall.

I've just got the error right now, and despite the stored value of `{"height":200,"width":160}` , when I double click, the size is not that at all. I think it's window 7 that is shrinking the window when it opens it outside the viewport.

**Comment by cxw42** [comment]
Related to #111 and #141, but neither is exactly the same as this one :) .  I'll tag it.

Summon is:

https://github.com/cxw42/TabFern/blob/27d8b2d71f59f34e7fcc7320162b77c10d044e98/app/bg/background.js#L61-L83

So, yes, summon does currently clip size based on the Chrome window to which you summon it.

So the backed-up position is (3,2) --- where did the window appear?  If you do a `chrome.windows.get` when the TF window is offscreen, what size/position do you get for the TF window?

**Comment by Procyon-b** [comment]
I'll check `chrome.windows.get` as soon as I get the error again.

Here is the procedure I always follow when closing chrome:

I close all windows one at a time, always finishing with my first/main window (if it has pined tabs).
Then the only window opened is TF; Sometimes minimized at the time (but I "restore" it); and I close it.
I close TF last to prevent "Recovered tabs" from appearing.

The solution I had in mind for some time now, is to prevent TF stroring new values if "left" and/or "top" are way off in the negative value.

**Comment by Procyon-b** [comment]
```
alwaysOnTop: false
focused: false
height: 200
id: 1084414374
incognito: false
left: -32000
state: "normal"
top: -32000
type: "popup"
width: 160
```

**Comment by cxw42** [comment]
That `-32000` is definitely not right :) .

I had one thought - in two places, we `getWindowSize(window)`:

https://github.com/cxw42/TabFern/blob/27d8b2d71f59f34e7fcc7320162b77c10d044e98/app/win/main_tl.js#L2953-L2981

In that code, `window` is `static/win/main.html`.  What we actually care about is the size of the container --- `static/win/container.html`.  I wonder if we are sometimes getting a position/size of `main` _with respect to `container`_ instead of with respect to the screen?

I'm not sure how to test that other than logging, but I am interested to know if you have any thoughts.

**Comment by cxw42** [comment]
> The solution I had in mind for some time now, is to prevent TF storing new values if "left" and/or "top" are way off in the negative value.

We can probably do that for `-32000` :) .  But eventually monitors will be that large, and the code will break!  I am hoping we can figure out the root cause of the `-32000`.

If worst comes to worst, we can make a profiling version of TF and run a timer to check every frame whether the numbers have gone wonky.  That would get us closer (at a rather severe cost).

**Comment by Procyon-b** [comment]
> In that code, window is static/win/main.html. What we actually care about is the size of the container --- static/win/container.html. I wonder if we are sometimes getting a position/size of main with respect to container instead of with respect to the screen?

I've checked.
When TF windows is minimized,
```js
chrome.storage.local.get( "tabfern-window-location", function(items){
  console.info(items);
  console.info( JSON.stringify(items) );
});
```
consistently gives `{"tabfern-window-location":{"height":27,"left":-32000,"top":-32000,"width":160}}`, and once restored, `{"tabfern-window-location":{"height":750,"left":3,"top":2,"width":281}}`

**Comment by cxw42** [comment]
Ah!  Yes, indeed --- that's a leftover from Win95, as I recall (nope - WinNT - [Old New Thing](https://devblogs.microsoft.com/oldnewthing/20041028-00/?p=37453)).  I see also [this SO question](https://stackoverflow.com/q/31755601/2877364) about the same.  Well, whaddya know!

### Confirmed!

With the developer console set to `tree-iframe (main.html)`, and the TF window visible:

    > chrome.windows.get(my_winid, (data)=>{console.log(data);});
    alwaysOnTop: false focused: false height: 535 id: 25 incognito: false left: 1527 
        state: "normal" top: 500 type: "popup" width: 383 __proto__: Object

After minimizing TF:

    > chrome.windows.get(my_winid, (data)=>{console.log(data);});
    alwaysOnTop: false focused: false height: 535 id: 25 incognito: false left: 1527 
        state: "minimized" top: 500 type: "popup" width: 383 __proto__: Object

Now, shortly after that, I got ---

    main.js:37129 Saved size

because the resize-detection timer expired and `timedResizeDetector()` ran.  I checked the resulting saved state and saw what you saw:

    > chrome.storage.local.get(K.LOCN_KEY,(data)=>{console.log(data);})
    tabfern-window-location: {height: 24, left: -32000, top: -32000, width: 160} __proto__: Object

I then restored the TF window and:

    main.js:37129 Saved size
    > chrome.storage.local.get(K.LOCN_KEY,(data)=>{console.log(data);})
    tabfern-window-location: {height: 535, left: 1527, top: 500, width: 383} __proto__: Object

**Comment by cxw42** [comment]
Please see [updated contribution guidelines](https://github.com/cxw42/TabFern/blob/master/CONTRIBUTING.md#Legal).  Thanks!

**Comment by Procyon-b** [comment]
I have tried a fix on a clone of the current version 0.2.0.1340. It works.
The fix relies only on verifying the value `.state == 'normal'` (from `chrome.windows.get(my_winid, ...` ).
`eventOnResize(evt)` and `timedResizeDetector()` still get their values from `getWindowSize(window)`.

Question:
Can we retire `getWindowSize()` entirely?
`basicInit()` would also need to be modified.
`chrome.windows.get()` gives us the necessary values. The only difference is that this function is async. Not a problem with the 2 timer/event based functions above, but I haven't looked at `basicInit()` yet, and how it is used.

**Side note**
I'll look at the guidelines on how to install the current code.
I'm on a win7/32. Is it a problem?
I know that github desktop is only 64bits. I suppose that command-line git is compatible, and other utils too.

**Comment by cxw42** [comment]
Win7 32 should be fine - windows git and Cygwin both still support x86, as far as I know.

I think using only `chrome.windows.get` would be fine for this.  The async has only been an issue for me in the past in drag-and-drop code.

Thank you!
